File: api/routers/admin_god_mode.py
"""
Admin God-Mode Control API
Comprehensive system management and emergency controls
"""
from fastapi import APIRouter, HTTPException, Depends
from typing import List, Dict, Any, Optional
from datetime import datetime
from pydantic import BaseModel
import logging
import asyncio
import json
try:
    import psutil
    PSUTIL_AVAILABLE = True
except ImportError:
    PSUTIL_AVAILABLE = False
    
try:
    import docker
    DOCKER_AVAILABLE = True
except ImportError:
    DOCKER_AVAILABLE = False

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/admin/api", tags=["god-mode-control"])

# Docker client (lazy init)
_docker_client = None

def get_docker_client():
    """Get Docker client with graceful fallback if docker module unavailable"""
    global _docker_client
    if not DOCKER_AVAILABLE:
        raise HTTPException(
            status_code=503,
            detail="Docker module not available. Install: pip install docker"
        )
    if _docker_client is None:
        try:
            # Try Docker socket connection
            _docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
        except Exception:
            # Fallback to environment
            _docker_client = docker.from_env()
    return _docker_client

# Models
class ServiceStatus(BaseModel):
    name: str
    status: str  # healthy, degraded, unhealthy
    cpu: float
    memory: float
    requests: int

class ServicesStatusResponse(BaseModel):
    services: List[ServiceStatus]
    timestamp: datetime

class SystemMetrics(BaseModel):
    system: Dict[str, Any]
    throughput: Dict[str, Any]
    timestamp: datetime

class BackfillRequest(BaseModel):
    years: int = 1
    max_symbols: Optional[int] = None

class RiskLimitsUpdate(BaseModel):
    max_position: float
    max_daily_loss: float
    max_leverage: float
    var_limit: float

class ScaleServiceRequest(BaseModel):
    replicas: int

# Service management endpoints
@router.get("/services/status")
async def get_services_status() -> ServicesStatusResponse:
    """
    Get status of all Docker services with resource metrics
    """
    try:
        client = get_docker_client()
        containers = client.containers.list()
        
        services = []
        for container in containers:
            # Get container stats (non-blocking)
            stats = container.stats(stream=False)
            
            # Calculate CPU percentage
            cpu_delta = stats['cpu_stats']['cpu_usage']['total_usage'] - \
                       stats['precpu_stats']['cpu_usage']['total_usage']
            system_delta = stats['cpu_stats']['system_cpu_usage'] - \
                          stats['precpu_stats']['system_cpu_usage']
            cpu_percent = (cpu_delta / system_delta) * 100.0 if system_delta > 0 else 0.0
            
            # Calculate memory usage in MB
            memory_mb = stats['memory_stats']['usage'] / (1024 * 1024)
            
            # Determine status based on health
            health = container.attrs.get('State', {}).get('Health', {})
            status_str = health.get('Status', 'unknown')
            
            if status_str == 'healthy':
                status = 'healthy'
            elif status_str in ['starting', 'unhealthy']:
                status = 'degraded' if container.status == 'running' else 'unhealthy'
            else:
                status = 'healthy' if container.status == 'running' else 'unhealthy'
            
            services.append(ServiceStatus(
                name=container.name,
                status=status,
                cpu=round(cpu_percent, 2),
                memory=round(memory_mb, 2),
                requests=0  # TODO: Pull from metrics endpoint if available
            ))
        
        return ServicesStatusResponse(
            services=services,
            timestamp=datetime.utcnow()
        )
        
    except Exception as e:
        logger.error(f"Failed to get services status: {e}")
        raise HTTPException(status_code=500, detail=f"Status check failed: {str(e)}")

@router.post("/services/{service_name}/restart")
async def restart_service(service_name: str):
    """
    Restart a specific service
    """
    try:
        client = get_docker_client()
        container = client.containers.get(service_name)
        container.restart(timeout=10)
        
        logger.info(f"Service {service_name} restart initiated")
        return {"status": "restarting", "service": service_name}
        
    except docker.errors.NotFound:
        raise HTTPException(status_code=404, detail=f"Service {service_name} not found")
    except Exception as e:
        logger.error(f"Failed to restart {service_name}: {e}")
        raise HTTPException(status_code=500, detail=f"Restart failed: {str(e)}")

@router.post("/services/{service_name}/scale")
async def scale_service(service_name: str, request: ScaleServiceRequest):
    """
    Scale a service to specified number of replicas
    """
    try:
        # TODO: Implement with Docker Swarm or Kubernetes API
        logger.info(f"Scaling {service_name} to {request.replicas} replicas")
        return {
            "status": "scaling",
            "service": service_name,
            "replicas": request.replicas
        }
        
    except Exception as e:
        logger.error(f"Failed to scale {service_name}: {e}")
        raise HTTPException(status_code=500, detail=f"Scaling failed: {str(e)}")

# System metrics streaming
@router.get("/metrics/stream")
async def stream_metrics():
    """
    Server-Sent Events stream of real-time system metrics
    """
    from fastapi.responses import StreamingResponse
    
    async def generate():
        while True:
            # Collect system metrics (with graceful fallback)
            if PSUTIL_AVAILABLE:
                cpu_percent = psutil.cpu_percent(interval=1)
                memory = psutil.virtual_memory()
                disk = psutil.disk_usage('/')
                net_io = psutil.net_io_counters()
                
                metrics = {
                    "system": {
                        "cpu": round(cpu_percent, 2),
                        "memory": round(memory.used / (1024**3), 2),  # GB
                        "disk": round(disk.percent, 2),
                        "network_mbps": round(net_io.bytes_sent / (1024**2), 2)
                    },
                    "throughput": {
                        "requests": 0,  # TODO: Pull from metrics
                        "data_mb": 0,
                        "signals": 0
                    }
                }
            else:
                # Fallback metrics when psutil unavailable
                metrics = {
                    "system": {
                        "cpu": 0,
                        "memory": 0,
                        "disk": 0,
                        "network_mbps": 0
                    },
                    "throughput": {
                        "requests": 0,
                        "data_mb": 0,
                        "signals": 0
                    }
                }
            
            yield f"event: metrics\ndata: {json.dumps(metrics)}\n\n"
            await asyncio.sleep(2)
    
    import json
    return StreamingResponse(generate(), media_type="text/event-stream")

# ML Model management
@router.post("/models/promote-all-shadows")
async def promote_all_shadows():
    """
    Promote all shadow models to production
    """
    try:
        # TODO: Integrate with ML governance service
        logger.info("Promoting all shadow models to production")
        return {"status": "promoted", "promoted": 0, "message": "Shadow promotion logic pending integration"}
        
    except Exception as e:
        logger.error(f"Shadow promotion failed: {e}")
        raise HTTPException(status_code=500, detail=f"Promotion failed: {str(e)}")

@router.post("/models/force-reload")
async def force_model_reload():
    """
    Force reload all ML models
    """
    try:
        # TODO: Send signal to ML inference service
        logger.info("Force reloading all ML models")
        return {"status": "reloading", "message": "Model reload initiated"}
        
    except Exception as e:
        logger.error(f"Model reload failed: {e}")
        raise HTTPException(status_code=500, detail=f"Reload failed: {str(e)}")

@router.get("/ml/models")
async def get_ml_models():
    """
    Get list of ML models with current weights
    """
    # TODO: Pull from ML config
    models = [
        {"id": "qwen2.5:72b", "name": "Qwen 2.5 72B", "weight": 0.25},
        {"id": "mixtral:8x22b", "name": "Mixtral 8x22B", "weight": 0.25},
        {"id": "command-r-plus:104b", "name": "Command R+ 104B", "weight": 0.20},
        {"id": "llama3.1:70b", "name": "Llama 3.1 70B", "weight": 0.15},
        {"id": "yi:34b", "name": "Yi 34B", "weight": 0.10},
        {"id": "phi3:14b", "name": "Phi-3 14B", "weight": 0.05}
    ]
    
    return {"models": models}

@router.put("/ml/models/{model_id}/weight")
async def update_model_weight(model_id: str, weight: float):
    """
    Update ensemble weight for a specific model
    """
    # TODO: Persist to ML config
    logger.info(f"Updating {model_id} weight to {weight}")
    return {"status": "updated", "model_id": model_id, "weight": weight}

# Factor management
@router.get("/factors/list")
async def get_factors():
    """
    Get list of all factors with current configuration
    """
    # TODO: Pull from factor engine config
    factors = [
        {"id": "momentum", "name": "Momentum", "enabled": True, "multiplier": 1.0},
        {"id": "value", "name": "Value", "enabled": True, "multiplier": 1.0},
        {"id": "quality", "name": "Quality", "enabled": True, "multiplier": 1.2},
        {"id": "growth", "name": "Growth", "enabled": True, "multiplier": 1.0},
        {"id": "volatility", "name": "Volatility", "enabled": True, "multiplier": 0.8},
        {"id": "size", "name": "Size", "enabled": False, "multiplier": 1.0}
    ]
    
    return {"factors": factors}

@router.post("/factors/{factor_id}/toggle")
async def toggle_factor(factor_id: str):
    """
    Toggle factor enabled/disabled
    """
    # TODO: Update factor engine config
    logger.info(f"Toggling factor {factor_id}")
    return {"status": "toggled", "factor_id": factor_id}

@router.put("/factors/{factor_id}/multiplier")
async def update_factor_multiplier(factor_id: str, multiplier: float):
    """
    Update factor multiplier
    """
    # TODO: Update factor engine config
    logger.info(f"Updating {factor_id} multiplier to {multiplier}")
    return {"status": "updated", "factor_id": factor_id, "multiplier": multiplier}

# Risk management
@router.post("/risk/update-limits")
async def update_risk_limits(limits: RiskLimitsUpdate):
    """
    Update system-wide risk limits
    """
    try:
        # TODO: Persist to risk management service
        logger.info(f"Updating risk limits: {limits}")
        return {"status": "updated", "limits": limits.dict()}
        
    except Exception as e:
        logger.error(f"Risk limits update failed: {e}")
        raise HTTPException(status_code=500, detail=f"Update failed: {str(e)}")

# Backfill controls
@router.post("/backfill/{data_type}")
async def trigger_backfill(data_type: str, request: BackfillRequest):
    """
    Trigger manual backfill for specified data type
    """
    try:
        # TODO: Send message to data-ingestion service
        logger.info(f"Triggering {data_type} backfill: {request.years} years, {request.max_symbols} symbols")
        return {
            "status": "scheduled",
            "data_type": data_type,
            "years": request.years,
            "max_symbols": request.max_symbols
        }
        
    except Exception as e:
        logger.error(f"Backfill trigger failed: {e}")
        raise HTTPException(status_code=500, detail=f"Backfill failed: {str(e)}")

# Emergency controls
@router.post("/emergency/kill-switch")
async def activate_kill_switch():
    """
    🚨 EMERGENCY: Halt all trading activity
    """
    try:
        # TODO: Send emergency stop signal to execution service
        logger.critical("🚨 KILL SWITCH ACTIVATED - Halting all trading")
        return {"status": "activated", "message": "All trading halted"}
        
    except Exception as e:
        logger.error(f"Kill switch failed: {e}")
        raise HTTPException(status_code=500, detail=f"Kill switch failed: {str(e)}")

@router.post("/circuit-breakers/reset")
async def reset_circuit_breakers():
    """
    Reset all circuit breakers
    """
    try:
        # TODO: Reset circuit breaker states
        logger.info("Resetting all circuit breakers")
        return {"status": "reset", "message": "Circuit breakers reset"}
        
    except Exception as e:
        logger.error(f"Circuit breaker reset failed: {e}")
        raise HTTPException(status_code=500, detail=f"Reset failed: {str(e)}")

@router.post("/system/force-gc")
async def force_garbage_collection():
    """
    Force garbage collection across services
    """
    try:
        import gc
        gc.collect()
        logger.info("Forced garbage collection")
        return {"status": "completed", "message": "Garbage collection executed"}
        
    except Exception as e:
        logger.error(f"GC failed: {e}")
        raise HTTPException(status_code=500, detail=f"GC failed: {str(e)}")

@router.post("/intelligence/save-config")
async def save_intelligence_config(config: Dict[str, Any]):
    """
    Save intelligence configuration (model weights, factors)
    """
    try:
        # TODO: Persist to config store
        logger.info("Saving intelligence configuration")
        return {"status": "saved", "message": "Configuration saved"}
        
    except Exception as e:
        logger.error(f"Config save failed: {e}")
        raise HTTPException(status_code=500, detail=f"Save failed: {str(e)}")

# Add to main API router
def register_god_mode_routes(app):
    """Register god-mode control routes with main app"""
    app.include_router(router)
    logger.info("God-mode control API registered")

File: api/static/js/admin_god_mode.js
// Admin God-Mode Control Panel
// Comprehensive system management and intelligence tweaking

class SystemControlPanel {
    constructor() {
        this.services = [];
        this.metrics = {};
        this.init();
    }

    init() {
        this.loadServices();
        this.setupRealtimeMonitoring();
        this.bindControls();
        this.startHealthChecks();
    }

    async loadServices() {
        try {
            const response = await fetch('/admin/api/services/status');
            const data = await response.json();
            this.services = data.services;
            this.renderServices();
        } catch (error) {
            console.error('Failed to load services:', error);
        }
    }

    renderServices() {
        const container = document.getElementById('services-grid');
        if (!container) return;
        
        const html = `
            <div class="services-control-grid">
                ${this.services.map(service => `
                    <div class="service-card ${service.status}">
                        <div class="service-header">
                            <h4>${service.name}</h4>
                            <span class="status-badge ${service.status}">${service.status}</span>
                        </div>
                        <div class="service-metrics">
                            <div class="metric">
                                <span class="label">CPU:</span>
                                <span class="value">${service.cpu}%</span>
                            </div>
                            <div class="metric">
                                <span class="label">Memory:</span>
                                <span class="value">${service.memory}MB</span>
                            </div>
                            <div class="metric">
                                <span class="label">Requests:</span>
                                <span class="value">${service.requests}/s</span>
                            </div>
                        </div>
                        <div class="service-controls">
                            <button class="btn-sm" onclick="systemControl.restartService('${service.name}')">Restart</button>
                            <button class="btn-sm" onclick="systemControl.viewLogs('${service.name}')">Logs</button>
                            <button class="btn-sm" onclick="systemControl.scaleService('${service.name}')">Scale</button>
                        </div>
                    </div>
                `).join('')}
            </div>
        `;
        
        container.innerHTML = html;
    }

    async restartService(serviceName) {
        if (!confirm(`Restart ${serviceName}?`)) return;
        
        try {
            await fetch(`/admin/api/services/${serviceName}/restart`, {
                method: 'POST',
                headers: this.getHeaders()
            });
            this.showNotification(`${serviceName} restart initiated`, 'success');
            setTimeout(() => this.loadServices(), 3000);
        } catch (error) {
            this.showNotification(`Restart failed: ${error.message}`, 'error');
        }
    }

    viewLogs(serviceName) {
        window.open(`/admin/logs/${serviceName}`, '_blank');
    }

    async scaleService(serviceName) {
        const replicas = prompt(`Number of replicas for ${serviceName}:`, '1');
        if (!replicas) return;
        
        try {
            await fetch(`/admin/api/services/${serviceName}/scale`, {
                method: 'POST',
                headers: this.getHeaders(),
                body: JSON.stringify({ replicas: parseInt(replicas) })
            });
            this.showNotification(`${serviceName} scaling to ${replicas} replicas`, 'success');
        } catch (error) {
            this.showNotification(`Scaling failed: ${error.message}`, 'error');
        }
    }

    setupRealtimeMonitoring() {
        const es = new EventSource('/admin/api/metrics/stream');
        es.addEventListener('metrics', (event) => {
            const metrics = JSON.parse(event.data);
            this.updateMetrics(metrics);
        });
    }

    updateMetrics(metrics) {
        this.metrics = metrics;
        
        // Update system-wide metrics
        this.updateElement('system-cpu', `${metrics.system.cpu}%`);
        this.updateElement('system-memory', `${metrics.system.memory}GB`);
        this.updateElement('system-disk', `${metrics.system.disk}%`);
        this.updateElement('system-network', `${metrics.system.network_mbps}Mbps`);
        
        // Update throughput
        this.updateElement('requests-per-sec', metrics.throughput.requests);
        this.updateElement('data-ingested', `${(metrics.throughput.data_mb / 1024).toFixed(2)}GB/hr`);
        this.updateElement('signals-generated', metrics.throughput.signals);
    }

    updateElement(id, value) {
        const el = document.getElementById(id);
        if (el) el.textContent = value;
    }

    bindControls() {
        // ML Model Controls
        this.bindButton('promote-all-shadows', () => this.promoteAllShadows());
        this.bindButton('force-model-reload', () => this.forceModelReload());
        this.bindButton('adjust-risk-limits', () => this.showRiskAdjuster());
        
        // Backfill Controls
        this.bindButton('trigger-equity-backfill', () => this.triggerBackfill('equity'));
        this.bindButton('trigger-news-backfill', () => this.triggerBackfill('news'));
        this.bindButton('trigger-options-backfill', () => this.triggerBackfill('options'));
        this.bindButton('trigger-social-backfill', () => this.triggerBackfill('social'));
        
        // System Controls
        this.bindButton('kill-switch', () => this.activateKillSwitch());
        this.bindButton('reset-circuit-breakers', () => this.resetCircuitBreakers());
        this.bindButton('force-gc', () => this.forceGarbageCollection());
    }

    bindButton(id, handler) {
        const btn = document.getElementById(id);
        if (btn) btn.addEventListener('click', handler);
    }

    async promoteAllShadows() {
        if (!confirm('Promote ALL shadow models to production?')) return;
        
        try {
            const response = await fetch('/admin/api/models/promote-all-shadows', {
                method: 'POST',
                headers: this.getHeaders()
            });
            const result = await response.json();
            this.showNotification(`Promoted ${result.promoted} models`, 'success');
        } catch (error) {
            this.showNotification(`Promotion failed: ${error.message}`, 'error');
        }
    }

    async forceModelReload() {
        if (!confirm('Force reload all ML models?')) return;
        
        try {
            await fetch('/admin/api/models/force-reload', {
                method: 'POST',
                headers: this.getHeaders()
            });
            this.showNotification('Model reload initiated', 'success');
        } catch (error) {
            this.showNotification(`Reload failed: ${error.message}`, 'error');
        }
    }

    showRiskAdjuster() {
        const modal = document.createElement('div');
        modal.className = 'modal-overlay';
        modal.innerHTML = `
            <div class="modal-content">
                <h2>Adjust Risk Limits</h2>
                <form id="risk-form">
                    <label>Max Position Size ($): <input type="number" name="max_position" value="100000"></label>
                    <label>Max Daily Loss ($): <input type="number" name="max_daily_loss" value="10000"></label>
                    <label>Max Leverage: <input type="number" step="0.1" name="max_leverage" value="2.0"></label>
                    <label>VaR Limit (95%): <input type="number" step="0.01" name="var_limit" value="0.05"></label>
                    <div class="modal-actions">
                        <button type="submit" class="button">Apply</button>
                        <button type="button" class="button secondary" onclick="this.closest('.modal-overlay').remove()">Cancel</button>
                    </div>
                </form>
            </div>
        `;
        
        document.body.appendChild(modal);
        
        modal.querySelector('form').addEventListener('submit', async (e) => {
            e.preventDefault();
            const formData = new FormData(e.target);
            const limits = Object.fromEntries(formData);
            
            try {
                await fetch('/admin/api/risk/update-limits', {
                    method: 'POST',
                    headers: this.getHeaders(),
                    body: JSON.stringify(limits)
                });
                this.showNotification('Risk limits updated', 'success');
                modal.remove();
            } catch (error) {
                this.showNotification(`Update failed: ${error.message}`, 'error');
            }
        });
    }

    async triggerBackfill(dataType) {
        const years = prompt(`Backfill ${dataType} data for how many years?`, dataType === 'equity' ? '20' : '5');
        if (!years) return;
        
        const symbols = prompt('Number of symbols (blank = all from watchlist):', '2000');
        
        try {
            await fetch(`/admin/api/backfill/${dataType}`, {
                method: 'POST',
                headers: this.getHeaders(),
                body: JSON.stringify({
                    years: parseInt(years),
                    max_symbols: symbols ? parseInt(symbols) : null
                })
            });
            this.showNotification(`${dataType} backfill started (${years}yr)`, 'success');
        } catch (error) {
            this.showNotification(`Backfill failed: ${error.message}`, 'error');
        }
    }

    async activateKillSwitch() {
        const confirmation = prompt('Type "KILL SWITCH" to halt all trading:');
        if (confirmation !== 'KILL SWITCH') return;
        
        try {
            await fetch('/admin/api/emergency/kill-switch', {
                method: 'POST',
                headers: this.getHeaders()
            });
            this.showNotification('🚨 KILL SWITCH ACTIVATED - All trading halted', 'error');
        } catch (error) {
            this.showNotification(`Kill switch failed: ${error.message}`, 'error');
        }
    }

    async resetCircuitBreakers() {
        try {
            await fetch('/admin/api/circuit-breakers/reset', {
                method: 'POST',
                headers: this.getHeaders()
            });
            this.showNotification('Circuit breakers reset', 'success');
        } catch (error) {
            this.showNotification(`Reset failed: ${error.message}`, 'error');
        }
    }

    async forceGarbageCollection() {
        try {
            await fetch('/admin/api/system/force-gc', {
                method: 'POST',
                headers: this.getHeaders()
            });
            this.showNotification('Garbage collection triggered', 'success');
        } catch (error) {
            this.showNotification(`GC failed: ${error.message}`, 'error');
        }
    }

    startHealthChecks() {
        setInterval(() => {
            this.checkSystemHealth();
        }, 5000);
    }

    async checkSystemHealth() {
        try {
            const response = await fetch('/health/full');
            const health = await response.json();
            
            // Update health indicators
            document.querySelectorAll('[data-health-component]').forEach(el => {
                const component = el.dataset.healthComponent;
                const status = health.components[component]?.status || 'unknown';
                el.className = `health-indicator ${status}`;
            });
        } catch (error) {
            console.error('Health check failed:', error);
        }
    }

    getHeaders() {
        const csrf = document.cookie.split(';').find(c => c.trim().startsWith('csrf_token='));
        return {
            'Content-Type': 'application/json',
            'X-CSRF-Token': csrf ? decodeURIComponent(csrf.split('=')[1]) : ''
        };
    }

    showNotification(message, type = 'info') {
        const notification = document.createElement('div');
        notification.className = `notification ${type}`;
        notification.textContent = message;
        document.body.appendChild(notification);
        
        setTimeout(() => {
            notification.classList.add('fade-out');
            setTimeout(() => notification.remove(), 500);
        }, 3000);
    }
}

// Intelligence Configuration Panel
class IntelligenceConfigurator {
    constructor() {
        this.models = [];
        this.factors = [];
        this.init();
    }

    async init() {
        await this.loadModels();
        await this.loadFactors();
        this.renderConfig();
    }

    async loadModels() {
        const response = await fetch('/admin/api/ml/models');
        const data = await response.json();
        this.models = data.models;
    }

    async loadFactors() {
        const response = await fetch('/admin/api/factors/list');
        const data = await response.json();
        this.factors = data.factors;
    }

    renderConfig() {
        const container = document.getElementById('intelligence-config');
        if (!container) return;
        
        const html = `
            <div class="config-panel">
                <h3>Model Weights & Ensembles</h3>
                <div class="model-weights">
                    ${this.models.map(model => `
                        <div class="weight-control">
                            <label>${model.name}</label>
                            <input type="range" min="0" max="1" step="0.01" value="${model.weight}"
                                   onchange="intelligenceConfig.updateModelWeight('${model.id}', this.value)">
                            <span class="weight-value">${(model.weight * 100).toFixed(0)}%</span>
                        </div>
                    `).join('')}
                </div>
                
                <h3>Factor Exposures</h3>
                <div class="factor-controls">
                    ${this.factors.map(factor => `
                        <div class="factor-control">
                            <label>${factor.name}</label>
                            <button class="btn-sm ${factor.enabled ? 'active' : ''}"
                                    onclick="intelligenceConfig.toggleFactor('${factor.id}')">
                                ${factor.enabled ? 'Enabled' : 'Disabled'}
                            </button>
                            <input type="number" step="0.1" value="${factor.multiplier}"
                                   onchange="intelligenceConfig.updateFactorMultiplier('${factor.id}', this.value)">
                        </div>
                    `).join('')}
                </div>
                
                <button class="button primary" onclick="intelligenceConfig.saveConfig()">
                    Save Configuration
                </button>
            </div>
        `;
        
        container.innerHTML = html;
    }

    async updateModelWeight(modelId, weight) {
        await fetch(`/admin/api/ml/models/${modelId}/weight`, {
            method: 'PUT',
            headers: window.systemControl.getHeaders(),
            body: JSON.stringify({ weight: parseFloat(weight) })
        });
    }

    async toggleFactor(factorId) {
        const factor = this.factors.find(f => f.id === factorId);
        factor.enabled = !factor.enabled;
        
        await fetch(`/admin/api/factors/${factorId}/toggle`, {
            method: 'POST',
            headers: window.systemControl.getHeaders()
        });
        
        this.renderConfig();
    }

    async updateFactorMultiplier(factorId, multiplier) {
        await fetch(`/admin/api/factors/${factorId}/multiplier`, {
            method: 'PUT',
            headers: window.systemControl.getHeaders(),
            body: JSON.stringify({ multiplier: parseFloat(multiplier) })
        });
    }

    async saveConfig() {
        await fetch('/admin/api/intelligence/save-config', {
            method: 'POST',
            headers: window.systemControl.getHeaders(),
            body: JSON.stringify({
                models: this.models,
                factors: this.factors
            })
        });
        
        window.systemControl.showNotification('Intelligence configuration saved', 'success');
    }
}

// Initialize God-Mode panels
window.addEventListener('DOMContentLoaded', () => {
    window.systemControl = new SystemControlPanel();
    window.intelligenceConfig = new IntelligenceConfigurator();
});
